extensions-core/s3-extensions/src/main/java/io/druid/storage/s3/S3DataSegmentMover.java

```diff
@@ -116,74 +117,22 @@ private void safeMove(
-                if (list.length == 0) {
-                  // should never happen
-                  throw new ISE("Unable to list object [s3://%s/%s]", s3Bucket, s3Path);
-                }
-                final S3Object s3Object = list[0];
-                if (s3Object.getStorageClass() != null &&
-                    s3Object.getStorageClass().equals(S3Object.STORAGE_CLASS_GLACIER)) {
-                  log.warn("Cannot move file[s3://%s/%s] of storage class glacier, skipping.", s3Bucket, s3Path);
-                } else {
-                  final String copyMsg = StringUtils.format(
-                      "[s3://%s/%s] to [s3://%s/%s]", s3Bucket,
-                      s3Path,
-                      targetS3Bucket,
-                      targetS3Path
-                  );
-                  log.info(
-                      "Moving file %s",
-                      copyMsg
-                  );
-                  final S3Object target = new S3Object(targetS3Path);
-                  if (!config.getDisableAcl()) {
-                    target.setAcl(GSAccessControlList.REST_CANNED_BUCKET_OWNER_FULL_CONTROL);
-                  }
-                  final Map<String, Object> copyResult = s3Client.moveObject(
-                      s3Bucket,
-                      s3Path,
-                      targetS3Bucket,
-                      target,
-                      false
-                  );
-                  if (copyResult != null && copyResult.containsKey("DeleteException")) {
-                    log.error("Error Deleting data after copy %s: %s", copyMsg, copyResult);
-                    // Maybe retry deleting here?
-                  } else {
-                    log.debug("Finished moving file %s", copyMsg);
-                  }
-                }
-              } else {
-                // ensure object exists in target location
-                if (s3Client.isObjectInBucket(targetS3Bucket, targetS3Path)) {
-                  log.info(
-                      "Not moving file [s3://%s/%s], already present in target location [s3://%s/%s]",
-                      s3Bucket, s3Path,
-                      targetS3Bucket, targetS3Path
-                  );
-                } else {
-                  throw new SegmentLoadingException(
-                      "Unable to move file [s3://%s/%s] to [s3://%s/%s], not present in either source or target location",
-                      s3Bucket,
-                      s3Path,
-                      targetS3Bucket,
-                      targetS3Path
-                  );
-                }
-              }
+          (Callable<Void>) () -> {
+            final String copyMsg = StringUtils.format(
+                "[s3://%s/%s] to [s3://%s/%s]",
+                s3Bucket,
+                s3Path,
+                targetS3Bucket,
+                targetS3Path
+            );
+            try {
+              selfCheckingMove(s3Bucket, targetS3Bucket, s3Path, targetS3Path, copyMsg);
```

Comment:
In addition to @gianm's comment, this can call ```copyObject()``` multiple times until the target object is seen even though the first ```copyObject()``` call succeeded. 

I think copying, checking the target object, and removing the source object should be done with separate retries.